python_package/madflow/makefile_template.py

```diff
@@ -0,0 +1,185 @@
+    """Adds C-Flags for CUDA (only if nvcc is found)"""
+    text = "ifeq ($(NVCC),)\n"
+    text += "CFLAGS_CUDA = $(CFLAGS)\n"
+    text += "CFLAGS_NVCC = ${TF_CFLAGS}\n"
+    text += "LDFLAGS_CUDA = $(LDFLAGS)\n"
+    text += "NVCC = $(CXX)\n"  # fallback in case nvcc is not installed
+    text += "else\n"
+    text += "CFLAGS_CUDA = $(CFLAGS) -D GOOGLE_CUDA=1 -I$(CUDA_PATH)/include\n"
+    text += (
+        "CFLAGS_NVCC = ${TF_CFLAGS} -O2 -std="
+        + cppVersion
+        + " -D GOOGLE_CUDA=1 -x cu -Xcompiler -fPIC -DNDEBUG --expt-relaxed-constexpr\n"
+    )
+    text += "LDFLAGS_CUDA = $(LDFLAGS) -L$(CUDA_PATH)/lib64 -lcudart\n"
+    text += "endif\n"
+    text += "\n"
+    return text
+
+
+def write_target():
+    """Adds the name of the generated library (matrix_processName_cu.so)"""
+    text = "TARGETS = $(shell ls gpu/ | grep \".h\" | sed 's/\.h/_cu.so/g')\n"
+    text += "\n"
+    return text
+
+
+def write_commands():
+    """Adds commands for compiling of source code"""
+    text = ""
+    text += write_generic_commands()
+    text += write_library_commands()
+    text += write_source_commands()
+    text += write_cleanup_commands()
+    return text
+
+
+def write_generic_commands():
+    """all compiles all target libraries (one for each subprocess, i.e.: qq~ -> X + gg -> X"""
+    text = "all: $(TARGETS)\n"
+    text += "\n"
+    return text
+
+
+def write_library_commands():
+    """Compile each library from object files"""
+    text = "%_cu.so: gpu/%.cudao gpu/%.cu.cudao\n"
+    text += "\t$(CXX) -o $@ $(CFLAGS_CUDA) $^ $(LDFLAGS_CUDA)\n"
+    text += "\n"
+    return text
+
+
+def write_source_commands():
+    """Generate object files"""
+    text = "%.o: %.cc\n"
+    text += "\t$(CXX) -c $(CFLAGS) $^ -o $@\n"
+    text += "\n"
+    text += "%.cu.cudao: %.cu.cc\n"
+    text += "\t$(NVCC) -c $(CFLAGS_NVCC) $^ -o $@\n"
+    text += "\n"
+    text += "%.cudao: %.cc\n"
+    text += "\t$(CXX) -c $(CFLAGS_CUDA) $^ -o $@\n"
+    text += "\n"
+    return text
+
+
+def write_cleanup_commands():
+    """Adds commmand for cleanup"""
+    # remove generated libraries
+    text = "clean:\n"
+    text += "\trm -f $(TARGETS) $(OBJECT_SRCS_CUDA)\n"
+    text += "\n"
+    # remove generated libraries and source code
+    text += "clean_all:\n"
+    text += "\trm -f $(TARGETS) $(OBJECT_SRCS_CUDA)\n"
+    text += "\trm -f gpu/*\n"
+    text += "\n"
+    return text
+
+
+def write_makefile(destination):
+
+    makefileContent = ""
+
+    makefileContent += write_compilers()
+    makefileContent += write_shell_name()
+    makefileContent += write_multithreading()
+    makefileContent += write_tf_generic_flags()
+    makefileContent += write_tf_cuda_flags()
+    # makefileContent += write_omp_flags()
+    makefileContent += write_cflags()
+    makefileContent += write_target()
+    makefileContent += write_commands()
+
+    # write the makefile
+    with open(destination + makefileName, "w") as fh:
+        fh.write(makefileContent)
```

Comment:
If you are already using paths you can do:

```suggestion
    (destination / makefile_name).write_text(makefile_content)
```

I would also appreciate the change from camelcase to underscore, so `makefileContent` -> `makefile_content` to follow the same style as the rest of the code ([pep8](https://www.python.org/dev/peps/pep-0008/)) but that's something that I can do myself automatically after the PR is merged.

Reply:
Done. Let me know if you still find camelcase names.